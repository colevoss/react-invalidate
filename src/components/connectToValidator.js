// @flow

import { Component, createElement } from 'react';
import invariant from 'invariant';
import type { CoreValidator } from '../shared/types';
import { CoreValidatorPropTypes } from '../shared/propTypes';

type ValidatorData = {
  isValid: boolean,
  validate: () => Promise<boolean>,
};

type State<T> = {
  mappedValidator: T,
};

type MapValidatorToProps<T> = (validatorData: ValidatorData, ownProps: Object) => T;

const connectToValidator = <ValidatorMap: Object>(mapValidatorToProps: MapValidatorToProps<ValidatorMap>) => {
  invariant(typeof mapValidatorToProps === 'function', 'mapValidatorToProps needs to be a function.');

  return (ConnectedComponent: ReactClass<ValidatorMap>): ReactClass<{}> => {
    class ConnectedToValidator extends Component {
      constructor(props, context) {
        super(props, context);

        invariant(
          context.validator,
          'connectToValidator needs to be used in the context of a <ValidationProvider> context.'
        );

        this.validator = context.validator;

        const mappedValidator = this.mapValidatorToProps();

        invariant(
          typeof mappedValidator === 'object',
          'provided mapValidatorToProps function needs to return an object.'
        );

        this.state = { mappedValidator };
      }

      componentDidMount() {
        this.unsubscribe = this.validator.subscribe(this.handleChange);
      }

      componentWillUnmount() {
        this.unsubscribe();
      }

      unsubscribe: Function;
      validator: CoreValidator;
      state: State<ValidatorMap>;

      handleChange = () => {
        this.setState({
          mappedValidator: this.mapValidatorToProps(),
        });
      };

      mapValidatorToProps(): ValidatorMap {
        return mapValidatorToProps(
          {
            isValid: this.validator.areAllValid(),
            validate: this.validator.validate,
          },
          this.props
        );
      }

      render() {
        return createElement(ConnectedComponent, this.state.mappedValidator);
      }
    }

    ConnectedToValidator.contextTypes = {
      validator: CoreValidatorPropTypes,
    };

    return ConnectedToValidator;
  };
};

export default connectToValidator;
