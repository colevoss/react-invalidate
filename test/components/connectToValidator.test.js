import React from 'react';
import { shallow, mount } from 'enzyme';
import connectToValidator from '../../src/components/connectToValidator';

const TestComp = props => <button {...props} />;

const mockCoreValidatorInitializer = (isValid = true) => {
  const mockUnsubscribe = jest.fn();
  const mockSubscribe = jest.fn(() => mockUnsubscribe);
  const mockValidate = jest.fn(() => Promise.resolve(isValid));

  return {
    subscribe: mockSubscribe,
    validate: mockValidate,
    areAllValid: () => isValid,
  };
};

describe('connectToValidator', () => {
  test('throws error if anything but a function is provided', () => {
    expect(() => connectToValidator()).toThrowErrorMatchingSnapshot();
  });

  test('throws error if mapValidatorToProps doesnt return object', () => {
    const WrappedTestComp = connectToValidator(() => {})(TestComp);

    const render = () => shallow(<WrappedTestComp className="test-comp" />, {
      context: {
        validator: mockCoreValidatorInitializer(),
      },
    });

    expect(render).toThrowErrorMatchingSnapshot();
  });

  test('returned function returns component as its given', () => {
    const WrappedTestComp = connectToValidator(() => ({}))(TestComp);
    const wrapper = shallow(<WrappedTestComp className="test-comp" />, {
      context: {
        validator: mockCoreValidatorInitializer(),
      },
    });

    expect(wrapper.is('.test-comp')).toBeTruthy;
  });

  test('subscribes to validator on mount with .handleChange', () => {
    const WrappedTestComp = connectToValidator(() => ({}))(TestComp);
    const mockCoreValidator = mockCoreValidatorInitializer();

    const wrapper = mount(<WrappedTestComp />, {
      context: {
        validator: mockCoreValidator,
      },
    });

    expect(mockCoreValidator.subscribe).toHaveBeenCalled();
    expect(mockCoreValidator.subscribe).toHaveBeenCalledWith(wrapper.instance().handleChange);
  });

  test('unsubscribes to validator on unmount', () => {
    const WrappedTestComp = connectToValidator(() => ({}))(TestComp);
    const mockCoreValidator = mockCoreValidatorInitializer();

    const wrapper = mount(<WrappedTestComp />, {
      context: {
        validator: mockCoreValidator,
      },
    });

    const spyOnUnsubscriber = jest.spyOn(wrapper.instance(), 'unsubscribe');

    wrapper.unmount();

    expect(spyOnUnsubscriber).toHaveBeenCalled();
  });

  test('sets state to the correct mappedState', () => {
    const mapValidatorToProps = ({ isValid }) => ({
      disabled: isValid,
    });

    const WrappedTestComp = connectToValidator(mapValidatorToProps)(TestComp);
    const mockCoreValidator = mockCoreValidatorInitializer();

    const wrapper = mount(<WrappedTestComp />, {
      context: {
        validator: mockCoreValidator,
      },
    });

    expect(wrapper.state().mappedValidator.disabled).toBeTruthy();
  });

  test('handleChange sets state to new mappedState', () => {
    const mapValidatorToProps = jest.fn();
    mapValidatorToProps.mockReturnValueOnce({
      disabled: true,
    });

    const WrappedTestComp = connectToValidator(mapValidatorToProps)(TestComp);
    const mockCoreValidator = mockCoreValidatorInitializer();

    const wrapper = mount(<WrappedTestComp />, {
      context: {
        validator: mockCoreValidator,
      },
    });

    expect(wrapper.state().mappedValidator.disabled).toBeTruthy();

    mapValidatorToProps.mockReturnValueOnce({
      disabled: false,
    });

    wrapper.instance().handleChange();

    expect(wrapper.state().mappedValidator.disabled).toBeFalsy();
  });
});
