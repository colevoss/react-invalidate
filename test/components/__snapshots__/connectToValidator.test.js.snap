// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`connectToValidator throws error if anything but a function is provided 1`] = `"mapValidatorToProps needs to be a function."`;

exports[`connectToValidator throws error if mapValidatorToProps doesnt return object 1`] = `"provided mapValidatorToProps function needs to return an object."`;
